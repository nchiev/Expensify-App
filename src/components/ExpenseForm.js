import React from 'react';
import moment from 'moment';
import  { SingleDatePicker } from 'react-dates';

//CSS
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
    //Default state of form
    state = {        
        isCalendarFocused: false,
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        error: ''
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;

        this.setState(() => {
            return {
                description: description
            };
        });
    };

    onNoteChange = (e) => {
        const note = e.target.value;

        this.setState(() => {
            return {
                note: note
            };
        });
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(amount.match(/^\d*(\.\d{0,2})?$/))
        {
            this.setState(() => ({amount}));        
        }
    };

    onDateChange = (createdAt) => {
        this.setState(() => {
            return {
                createdAt: createdAt
            }
        });
    };

    onDateFocusChange = (e) => {
        this.setState(() => {
            return {
                isCalendarFocused: e.focused
            }
        });
    };

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            this.setState(() => {
                return {
                    error: 'Please provide description and amount'
                }
            });
        } else {
            this.setState(() => {
                return {
                    error: ''
                }
            });
            this.props.onSubmit({
                    description: this.state.description,
                    amount: parseFloat(this.state.amount, 10) * 100,
                    createdAt: this.state.createdAt.valueOf(),
                    note: this.state.note
            });
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                     type="text"
                     placeholder="Description"
                     autoFocus
                     value={this.state.description}
                     onChange={this.onDescriptionChange}
                    />
                    <input 
                     type="text"
                     placeholder="Amount"
                     value={this.state.amount}
                     onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                     date={this.state.createdAt}
                     onDateChange={this.onDateChange} 
                     focused={this.state.isCalendarFocused} 
                     onFocusChange={this.onDateFocusChange} 
                     numberOfMonths={1}
                     isOutsideRange={() => false} //Select any day
                    />
                    <textarea
                     placeholder="Add your notes (Optional)"
                     value={this.state.note}
                     onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}