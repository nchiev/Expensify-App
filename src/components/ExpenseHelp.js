import React from 'react';

const ExpenseHelp = () => (
    <div>
        Help
    </div>
);

export default ExpenseHelp;