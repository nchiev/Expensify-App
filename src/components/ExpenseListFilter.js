import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import { setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate } from '../actions/filters'

class ExpenseListFilter extends React.Component {
    state = {
        calendarFocused: undefined
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }

    onFocusInput = (calendarFocused) => {
        this.setState(() => {
            return {
                calendarFocused
            }
        });
    }

    handleExpenseTextFilterChange = (event) => {
        this.props.dispatch(setTextFilter(event.target.value));
    };

    handleExpenseTypeChange = (event) => {
        const expenseTypeFilter = event.target.value;

        switch (expenseTypeFilter){
            case 'date':
                this.props.dispatch(sortByDate());
                break;
            case 'amount':
                this.props.dispatch(sortByAmount());
                break;
        }
    };

    render() {
        return (
            <div>
                <input type="text" onChange={(e)=> {
                    this.handleExpenseTextFilterChange(e);
                }}
                value={this.props.filters.text}/>
                <select onChange={(e)=> {
                        this.handleExpenseTypeChange(e)
                    }}
                    value={this.props.filters.sortBy}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                 startDate={this.props.filters.startDate}
                 endDate={this.props.filters.endDate}
                 onDatesChange={this.onDatesChange}
                 focusedInput={this.state.calendarFocused}
                 onFocusChange={this.onFocusInput} 
                 showClearDates={true}
                 numberOfMonths={1}
                 isOutsideRange={() => false}
                />
            </div>
        ); 
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

export default connect(mapStateToProps)(ExpenseListFilter);