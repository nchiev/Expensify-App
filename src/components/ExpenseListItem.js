import React from 'react';
import { connect } from 'react-redux';
import { Link } from "react-router-dom";

import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({ id, description, amount, createdAt, dispatch }) => {
    const handleDeleteExpense = (id ,dispatch) => {
        dispatch(removeExpense({ id }));
    };

    return (
        <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>{amount} - {createdAt}</p>
        
        <button onClick={ (e) => 
            handleDeleteExpense(id,dispatch) 
        }>Remove</button>
    </div>
    );
};

export default connect()(ExpenseListItem);;