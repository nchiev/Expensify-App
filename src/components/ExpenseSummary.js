import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';

import selectExpenses from '../selectors/expenses'
import getExpenseTotal from '../selectors/expenses-total'

export const ExpenseSummary = ({expenseCount, expenseTotal}) => {
    const expenseWord = expenseCount && expenseCount > 1 ? 'expenses' : 'expense'  ;
    const formattedTotal = numeral(expenseTotal/100).format('$0,0.00');
    expenseCount = expenseCount ? expenseCount : 0;
    
    return (
        <div>
            <p>{`Viewing ${expenseCount} ${expenseWord} totalling ${formattedTotal}`}</p>
        </div>
    );
}

const mapStateToProps = (state) => {
    const filteredExpense = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: filteredExpense.length,
        expenseTotal: getExpenseTotal(filteredExpense)
    };
};

export default connect(mapStateToProps)(ExpenseSummary);