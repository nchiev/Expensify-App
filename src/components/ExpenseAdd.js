import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expenses'

export class ExpenseAdd extends React.Component {
    onSubmit = (expense) => {
        this.props.onSubmit(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={ this.onSubmit }/>
            </div>
        );
    }
}

const mapDisptachToProps = (dispatch) => {
    return {
        onSubmit: (expense) => dispatch(startAddExpense(expense))
    };
};

export default connect(undefined, mapDisptachToProps)(ExpenseAdd);