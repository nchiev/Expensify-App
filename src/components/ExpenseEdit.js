import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm'
import { startEditExpense, startDeleteExpense } from '../actions/expenses'

export class ExpenseEdit extends React.Component {
    handleDeleteExpense = () => {
        this.props.handleDeleteExpense({ id : this.props.expense.id });
        this.props.history.push('/');
    };

    onSubmit = (expense) => {
        console.log(this.props.expense);
        this.props.onSubmit( this.props.expense.id, expense );
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm 
                expense={ this.props.expense } 
                onSubmit={ this.onSubmit }/>
                <button onClick={ this.handleDeleteExpense }>Remove</button>
            </div>
        );
    }
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
};

const mapDisptachToProps = (dispatch) => {
    return {
        onSubmit: (id, expense) => dispatch(startEditExpense(id, expense)),
        handleDeleteExpense: (id) => dispatch(startDeleteExpense(id))
    };
};

export default connect(mapStateToProps, mapDisptachToProps)(ExpenseEdit);