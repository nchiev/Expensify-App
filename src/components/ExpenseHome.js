import React from 'react';
import ExpenseList from './ExpenseList'
import ExpenseSummary from './ExpenseSummary';
import ExpenseListFilter from './ExpenseListFilter';


const ExpenseHome = () => (
    <div>
        <ExpenseListFilter />
        <ExpenseSummary />
        <ExpenseList />
    </div>
);

export default ExpenseHome;