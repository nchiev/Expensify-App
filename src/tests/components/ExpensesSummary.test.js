import React from 'react';
import { shallow } from 'enzyme';

import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should render no expenses', () => {
    const wrapper = shallow(<ExpenseSummary/>);

    expect(wrapper).toMatchSnapshot();
});

test('should render expense count with 12 and total $2.00', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={12} expenseTotal={200}/>);

    expect(wrapper).toMatchSnapshot();
});