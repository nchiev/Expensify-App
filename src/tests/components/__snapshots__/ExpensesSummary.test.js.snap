// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense count with 12 and total $2.00 1`] = `
<div>
  <p>
    Viewing 12 expenses totalling $2.00
  </p>
</div>
`;

exports[`should render no expenses 1`] = `
<div>
  <p>
    Viewing 0 expense totalling $0.00
  </p>
</div>
`;
